ext/random/random.c

```diff
@@ -0,0 +1,1754 @@
+		/* char (8 bit) * 32 = 256 bits */
+		if (str_seed->len == 32) {
+			/* Endianness safe copy */
+			int i;
+			char *ptr = (char *) &state->s;
+			for (i = 0; i < 32; i++) {
+				ptr[i] = str_seed->val[i];
+			}
+		}  else {
+			zend_argument_value_error(1, "state strings must be 32 bytes");
+			RETURN_THROWS();
+		}
+	} else {
+		engine->algo->seed(state, int_seed);
+	}
+}
+/* }}} */
+
+/* {{{ Construct object */
+PHP_METHOD(Random_Randomizer, __construct)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_object *engine_object = 0;
+	zval zengine_object;
+	zend_string *pname;
+	
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS_OR_NULL(engine_object, random_ce_Random_Engine);
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* Create default RNG instance */
+	if (!engine_object) {
+		engine_object = php_random_engine_secure_new(random_ce_Random_Engine_Secure);
+		zend_call_known_instance_method_with_0_params(
+			random_ce_Random_Engine_Secure->constructor,
+			engine_object,
+			NULL
+		);
+
+		/* No need self-refcount */
+		GC_DELREF(engine_object);
+	}
+
+	ZVAL_OBJ(&zengine_object, engine_object);
+
+	/* Write property */
+	pname = zend_string_init("engine", sizeof("engine") - 1, 0);
+	zend_std_write_property(&randomizer->std, pname, &zengine_object, NULL);
+	zend_string_release(pname);
+
+	php_random_randomizer_common_init(randomizer, engine_object);
+}
+/* }}} */
+
+/* {{{ Generate random number in range */
+PHP_METHOD(Random_Randomizer, getInt)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_long min, max;
+	int argc = ZEND_NUM_ARGS();
+
+	if (argc == 0) {
+		// right shift for compatibility
+		RETURN_LONG(((uint64_t) randomizer->algo->generate(randomizer->state)) >> 1);
+	}
+
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_LONG(min)
+		Z_PARAM_LONG(max)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (UNEXPECTED(max < min)) {
+		zend_argument_value_error(2, "must be greater than or equal to argument #1 ($min)");
+		RETURN_THROWS();
+	}
+
+	RETURN_LONG(php_random_engine_range(randomizer->algo, randomizer->state, min, max));
+}
+/* }}} */
+
+/* {{{ Generate random bytes string in ordered length */
+PHP_METHOD(Random_Randomizer, getBytes)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_string *retval;
+	zend_long length;
+	size_t generated_size = 0;
+	int i;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_LONG(length)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (length < 1) {
+		zend_argument_value_error(1, "must be greater than 0");
+		RETURN_THROWS();
+	}
+
+	retval = zend_string_alloc(length, 0);
+
+	while (generated_size < (length * sizeof(char))) {
+		uint64_t generated;
+		size_t generate_size;
+
+		RANDOM_ENGINE_GENERATE_SIZE(randomizer->algo, randomizer->state, generated, generate_size);
+
+		for (i = 0; i < generate_size; i++) {
+			ZSTR_VAL(retval)[i] = (generated >> (i * 8)) & 0xff;
+		}
+
+		generated_size += generate_size;
+	}
```

Comment:
This implementation is not correct:

1. You always override the first 4 / 8 bytes of the string, because you don't take the offset into account properly.
2. You will write into illegal memory, if the requested length is not a multiple of 4 / 8, because the inner loop does not check the `length`.

Reply:
fixed: [0b3d5ba](https://github.com/php/php-src/pull/8094/commits/0b3d5bac377868192d3f329576f4806294804aa5)